apps/aefate/src/aefa_fate.erl

```diff
@@ -374,18 +397,46 @@ push(V, ES) ->
 push_return_address(ES) ->
     aefa_engine_state:push_return_address(ES).
 
+%% Push a gas cap on the call stack to limit the available gas, but keep the
+%% remaining gas around.
+push_gas_cap(Gas, ES) when not ?IS_FATE_INTEGER(Gas) ->
+    abort({value_does_not_match_type, Gas, integer}, ES);
+push_gas_cap(Gas, ES) when ?IS_FATE_INTEGER(Gas) ->
+    GasInt = ?FATE_INTEGER_VALUE(Gas),
+    case GasInt > 0 of
+        false ->
+            abort({call_error, bad_gas_cap}, ES);
+        true ->
+            aefa_engine_state:push_gas_cap(GasInt, ES)
+    end.
+
 pop_call_stack(ES) ->
     case aefa_engine_state:call_stack(ES) of
         [] -> {stop, ES};
-        [{Contract, Function, BB, Mem}| Rest] ->
-            ES1 = aefa_engine_state:set_call_stack(Rest, ES),
-            %% The memory is functional and restored
-            %% to the version before the call.
+        [{gas_store, StoredGas}| Rest] ->
+            Gas = StoredGas + aefa_engine_state:gas(ES),
+            ES1 = aefa_engine_state:set_gas(Gas, ES),
+            ES2 = aefa_engine_state:set_call_stack(Rest, ES1),
+            pop_call_stack(ES2);
+        [{Contract, Function, BB, Mem, Value}| Rest] ->
+            ES0 = aefa_engine_state:set_call_value(Value, ES),
+            ES1 = aefa_engine_state:set_call_stack(Rest, ES0),
             ES2 = aefa_engine_state:set_memory(Mem, ES1),
             ES3 = set_function(Contract, Function, ES2),
             {jump, BB, ES3}
     end.
 
+abort_out_of_gas(ES) ->
```

Comment:
Doesn't actually abort, so call it something else?

Reply:
Sure. It used to do abort in the end ;-)